src/dragonball/src/device_manager/console_manager.rs

```diff
@@ -0,0 +1,427 @@
+            let events = Events::with_data(&conn_sock, EPOLL_EVENT_SERIAL_DATA, EventSet::IN);
+            if let Err(e) = ops.add(events) {
+                slog::error!(self.logger,
+                    "failed to register epoll event for serial, {:?}", e;
+                    "subsystem" => "console_mgr",
+                );
+                return Err(std::io::Error::last_os_error());
+            }
+
+            let conn_sock_copy = conn_sock.try_clone()?;
+            // Do not expected poisoned lock.
+            self.device
+                .lock()
+                .unwrap()
+                .set_output_stream(Some(Box::new(conn_sock_copy)));
+
+            self.sock_conn = Some(conn_sock);
+        }
+
+        Ok(())
+    }
+
+    fn uds_read_in(&mut self, ops: &mut EventOps) -> std::io::Result<()> {
+        let mut should_drop = true;
+
+        if let Some(conn_sock) = self.sock_conn.as_mut() {
+            let mut out = [0u8; 64];
+            match conn_sock.read(&mut out[..]) {
+                Ok(0) => {
+                    // Zero-length read means EOF. Remove this conn sock.
+                    self.device
+                        .lock()
+                        .expect("console: poisoned console lock")
+                        .set_output_stream(None);
+                }
+                Ok(count) => {
+                    self.device
+                        .lock()
+                        .expect("console: poisoned console lock")
+                        .raw_input(&out[..count])?;
+                    should_drop = false;
+                }
+                Err(e) => {
+                    slog::warn!(self.logger,
+                        "error while reading serial conn sock: {:?}", e;
+                        "subsystem" => "console_mgr"
+                    );
+                    self.device
+                        .lock()
+                        .expect("console: poisoned console lock")
+                        .set_output_stream(None);
+                }
+            }
+        }
+
+        if should_drop {
+            assert!(self.sock_conn.is_some());
+            // Safe to unwrap() because self.sock_conn is Some().
+            let sock_conn = self.sock_conn.take().unwrap();
+            let events = Events::with_data(&sock_conn, EPOLL_EVENT_SERIAL_DATA, EventSet::IN);
+            if let Err(e) = ops.remove(events) {
+                slog::error!(self.logger,
+                    "failed deregister epoll event for UDS, {:?}", e;
+                    "subsystem" => "console_mgr"
+                );
+            }
+        }
+
+        Ok(())
+    }
+
+    fn stdio_read_in(&mut self, ops: &mut EventOps) -> std::io::Result<()> {
+        let mut should_drop = true;
+
+        if self.stdin_handle.is_some() {
+            let mut out = [0u8; 64];
+            // Safe to unwrap() because self.stdin_handle is Some().
+            let stdin_lock = self.stdin_handle.as_ref().unwrap().lock();
```

Comment:
This would be better coded using `let ... if` to avoid the check-and-unwrap combination:

```rust
let handle = if let Some(self.stdin_handle) {
  // ...
}
```

Same comment for other `unwrap()` calls.

Reply:
Changed.